crypto/src/utils.rs

```diff
@@ -0,0 +1,205 @@
+// Copyright 2018 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![allow(dead_code)] // TODO Remove after complete ECR-2518 and  ECR-2519
+use super::{gen_keypair, gen_keypair_from_seed, PublicKey, SecretKey, Seed, SEED_LENGTH};
+use hex_buffer_serde::Hex;
+use pwbox::{sodium::Sodium, ErasedPwBox, Eraser, Suite};
+use rand::thread_rng;
+use std::borrow::Cow;
+#[cfg(unix)]
+use std::os::unix::fs::{MetadataExt, OpenOptionsExt};
+use std::{
+    fs::{File, OpenOptions},
+    io::{Error, ErrorKind, Read, Write},
+    path::Path,
+};
+use toml;
+
+/// Creates a TOML file that contains encrypted `SecretKey` and returns `PublicKey` for the secret key.
+pub fn create_keys_file<P: AsRef<Path>>(path: P, pass_phrase: &[u8]) -> Result<PublicKey, Error> {
+    let (pk, sk) = gen_keypair();
+    let keys = EncryptedKeys::encrypt(pk, &sk, pass_phrase)?;
+    let file_content =
+        toml::to_string_pretty(&keys).map_err(|e| Error::new(ErrorKind::Other, e))?;
+    let mut open_options = OpenOptions::new();
+    open_options.create(true).write(true);
+    #[cfg(unix)]
+    open_options.mode(0o600);
+    let mut file = open_options.open(path.as_ref())?;
+    file.write_all(file_content.as_bytes())?;
+
+    Ok(pk)
+}
+
+/// Reads `PublicKey` and `SecretKey` from encrypted file located by path and returns its.
+pub fn read_keys_from_file<P: AsRef<Path>>(
+    path: P,
+    pass_phrase: &[u8],
+) -> Result<(PublicKey, SecretKey), Error> {
+    let mut key_file = File::open(path)?;
+
+    if cfg!(unix) {
+        let file_info = key_file.metadata()?;
+        if (file_info.mode() & 0o600) != 0o600 {
```

Comment:
I'm pretty sure this comparison doesn't check what needs to be checked.

```rust
let mode: u32 = 0o664;
assert_eq!(mode & 0o600, 0o600);
```

I think the correct one is

```rust
(file_info.mode() & 0o077) = 0
```

(i.e., group and other bits are not set). Please add a test for this.